chronos/bipbuffer.nim

```diff
@@ -0,0 +1,167 @@
+## Cooke.
+##
+## The Bip-Buffer is like a circular buffer, but slightly different. Instead of
+## keeping one head and tail pointer to the data in the buffer, it maintains two
+## revolving regions, allowing for fast data access without having to worry
+## about wrapping at the end of the buffer. Buffer allocations are always
+## maintained as contiguous blocks, allowing the buffer to be used in a highly
+## efficient manner with API calls, and also reducing the amount of copying
+## which needs to be performed to put data into the buffer. Finally, a two-phase
+## allocation system allows the user to pessimistically reserve an area of
+## buffer space, and then trim back the buffer to commit to only the space which
+## was used.
+##
+## https://www.codeproject.com/Articles/3479/The-Bip-Buffer-The-Circular-Buffer-with-a-Twist
+
+{.push raises: [].}
+
+import results
+
+type
+  BipPos = object
+    start: Natural
+    finish: Natural
+
+  BipBuffer* = object
+    a, b, r: BipPos
+    data: seq[byte]
+
+proc init*(t: typedesc[BipBuffer], size: int): BipBuffer =
+  ## Creates new Bip Buffer with size `size`.
+  BipBuffer(data: newSeq[byte](size))
+
+template len(pos: BipPos): Natural =
+  pos.finish - pos.start
+
+template zero(pos: var BipPos) =
+  pos = BipPos()
+
+func init(t: typedesc[BipPos], start, finish: Natural): BipPos =
+  BipPos(start: start, finish: finish)
+
+func calcReserve(bp: BipBuffer): tuple[space: Natural, start: Natural] =
+  if len(bp.b) > 0:
+    (Natural(bp.a.start - bp.b.finish), bp.b.finish)
+  else:
+    let spaceAfterA = Natural(len(bp.data) - bp.a.finish)
+    if spaceAfterA >= bp.a.start:
+      (spaceAfterA, bp.a.finish)
+    else:
+      (bp.a.start, Natural(0))
+
+func freeSpace*(bp: BipBuffer): Natural =
+  ## Returns amount of free space in buffer `bp`.
+  var sum: Natural
+  if len(bp.b) > 0:
+    sum += (bp.a.start - bp.b.finish)
+  else:
+    sum += bp.a.start
+  sum += Natural(len(bp.data)) - bp.a.finish
+  sum
+
+func availSpace*(bp: BipBuffer): Natural =
+  ## Returns amount of space available for reserve in buffer `bp`.
+  let (res, _) = bp.calcReserve()
+  res
+
+func usedSpace*(bp: BipBuffer): Natural =
+  ## Returns amount of used space in buffer `bp`.
+  len(bp.b) + len(bp.a)
+
+proc reserve*(bp: var BipBuffer, size: Natural): Result[Natural, cstring] =
+  ## Reserve `size` bytes in buffer. Returns number of bytes reserved.
+  doAssert(size <= len(bp.data))
+  let (availableSpace, reserveStart) = bp.calcReserve()
+  if availableSpace == 0:
+    return err("Not enough space available")
+  let reserveLength = min(availableSpace, size)
+  bp.r = BipPos.init(reserveStart, Natural(reserveStart + reserveLength))
+  ok(availableSpace)
+
+proc reserve*(bp: var BipBuffer): Natural =
+  ## Reserve all available free space in buffer. Returns number of bytes
+  ## reserved.
+  ##
+  ## Note, this procedure will raise Defect if there is no free space available.
```

Comment:
it would be more natural to return a `Result` here, like the other `reserve` - callers that *know* there is space available can do `reserve().expect("free bytes")` - then it is clearly documented on the calling side what the expectation is

Reply:
basically, a single `reserve(bp: var BipBuffer, size = Natural.high()` is all that's needed

Reply:
Fixed in https://github.com/status-im/nim-chronos/pull/485/commits/c74f451a373e4049178e2841a2b20695b5cd5615